### PR TITLE
Workflow Forking docs

### PR DESCRIPTION
**Pull request by awholegunch**
Re-formats the "Building Block" Workflow page, adds two draft images, and adds documentation on the "fork" feature of Workflows.

Staged at ["Building Block" Workflow](http://stg-web101.sw.joyent.continuuity.net/cdap/2.8.0-SNAPSHOT-r2.8_workflow_forking_docs/en/developers-manual/building-blocks/workflows.html)

https://issues.cask.co/browse/CDAP-1839

### CODE REVIEW

**Comment by prinam** [comment]
@anew it'll be nice to have you look at this too, @sagarkapare is looking at this anwyay.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 1]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+   :align: center
+
+Forks and Joins
+---------------
+
+To create such a Workflow, you provide a series of *forks* and *joins* in your Workflow
+specification, following these rules:
+
+- Where your control flow initially splits, you place a ``fork`` method. 
+- Every time your control flow splits, you add additional ``.fork`` methods.
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: additional ``fork`` methods

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 2]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+
+    @Override
+    public void configure() {
+      setName("ParallelWorkflow");
+      setDescription("Demonstration of a parallel Workflow");
+      
+      addMapReduce("JoinWithCatalogMR");
+    
+      fork()
+        .addMapReduce("BuildProductProfileMR");
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: no `;` at the end

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 3]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+      .also()
+        .addMapReduce("BuildUserProfileMR");
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: no `;` required at the end

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 4]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+      
+      addAction(new UploadProfilesCA());
+    }
+  }
+
+Provided that the control flow does not have cycles or the joining of any branches that do
+not originate from the same fork, flows of different complexity can be created using these
+rules and methods.
+
+More complicated structures can be created using ``.fork``. To add another MapReduce
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: created using `fork`

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 5]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+
+- Where your control flow initially splits, you place a ``fork`` method. 
+- Every time your control flow splits, you add additional ``.fork`` methods. 
+- Every point where you have either a program or an action, you add a ``.addMapReduce``,
+  ``.addSpark``, or ``.addAction`` method. 
+- To show each fork, use a ``.also`` method to separate the different branches of the
+  control flow. 
+- Where your control flow reconnects, you add a ``.join`` method to indicate. 
+- The control flow always concludes with a ``.join`` method.
+
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: I think we do not need to have `.` before the method names.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: fixed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: fixed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 6]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+Concurrent Workflows
+--------------------
+By default, a Workflow runs sequentially. Multiple instances of a Workflow can be run
+concurrently. To enable concurrent runs for a Workflow, set its runtime argument
+``concurrent.runs.enabled`` to ``true``.
+
+.. _workflow_parallel:
+
+Parallel Workflows
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Should we call it ``Parallelizing Workflow execution``. The reason is ``Parallel Workflows`` makes it feel like we are executing the entire Workflow in parallel (which would be similar to the Concurrent Workflow above).

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 7]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+- The control flow always concludes with a ``join`` method.
+
+The application shown above could be coded (assuming the other classes referred to exist) as::
+
+  public class ParallelWorkflow extends AbstractWorkflow {
+
+    @Override
+    public void configure() {
+      setName("ParallelWorkflow");
+      setDescription("Demonstration of a parallel Workflow");
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: ``Demonstration of parallelizing execution of the Workflow``

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 8]
```diff
@@ -68,14 +83,111 @@ custom action, you will need to extend the ``AbstractWorkflowAction`` and implem
...
+
+More complicated structures can be created using ``fork``. To add another MapReduce
+that runs in parallel to the entire process described above, you could use code such as::
+
+  public class ParallelWorkflow extends AbstractWorkflow {
+
+    @Override
+    public void configure() {
+      setName("DoubleParallelWorkflow");
+      setDescription("Demonstration of a double parallel Workflow");
```
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: ``Demonstration of parallel execution using complex fork in the Workflow``

**Comment by sagarkapare** [comment]
Minor comments. Otherwise LGTM :+1:

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 6 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: works for me.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 7 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: done.

**Comment by awholegunch** [comment]
Edited and restaged.

**Pull request merged**